### PR TITLE
Story2 into main

### PR DESCRIPTION
**Pull request by Jago971**
I believe this ticks:

requirements - shows collection, can add to collection, displays additions
it is unit tested - tests appropriately named
it is accessible - it has labels
it is responsive - tested to 350px width

### CODE REVIEW

File: styles.css
Comment (DBaldwin8) [inline styles.css:1]: Suggestion: add white space to ease reading.

File: tests/dropDownOptionsTest.php
Comment (DBaldwin8) [inline tests/dropDownOptionsTest.php:1]: Suggestion: add white space between functions to ease reading.

**Review by DBaldwin8 — COMMENTED** [review]
Suggestions made for formatting.

File: index.php [hunk 1]
```diff
@@ -20,14 +20,44 @@
...
+                <input class="submit" type="submit">
+            </div>
+        </form>
     </div>
+    <div class="interaction">
+        <h1>EDIT</h1>
+    </div>
+    <div class="interaction">
+        <h1>DELETE</h1>
+    </div>
```
Comment (mallowmew) [inline index.php]: EDIT and DELETE are not part of this story, remove these.

We don't do work that isn't in the story - what if we don't get DELETE done and a big box is there on the page where it "will go", what will our client or users think? What awkward questions are they going to ask? Are they gong to accept the software with big obvious holes in it?

File: src/functions.php [hunk 2]
```diff
@@ -40,4 +42,56 @@ function displaySocksCollection(PDO $db): string {
...
+}
+
+function createDropdown(string $name) : string {
+    $start = "<select name=\"{$name}\" id=\"{$name}\"><option disabled=\"disabled\" selected=\"selected\">$name</option>";
+    $options = dropDownOptions(getTable($name, connectDB()));
+    $end = '</select>';
+    return $start . $options . $end;
+}
+
+if ($_GET) {
```
Comment (mallowmew) [inline src/functions.php]: wrong HTTP verb for ADD

File: src/functions.php [hunk 3]
```diff
@@ -40,4 +42,56 @@ function displaySocksCollection(PDO $db): string {
...
+    $query = $db->prepare("INSERT INTO `socks` (`name`, `size`, `pattern`, `color`, `description`) VALUES (:name, :size, :pattern, :color, :description)");
+    //$query = $db->prepare("INSERT INTO `socks` (`size`, `pattern`, `color`) VALUES (:size, :pattern, :color)");
+    $result = $query->execute([
+        'name' => $_GET['name'],
+        'size' => getRelatedNumberForDropdownOption('size', $_GET['size']),
+        'pattern' => getRelatedNumberForDropdownOption('pattern', $_GET['pattern']),
+        'color' => getRelatedNumberForDropdownOption('color', $_GET['color']),
+        'description' => $_GET['description']
+    ]);
+}
```
Comment (mallowmew) [inline src/functions.php]: This is not in a function - it only accidentally works because you `require` the file it is in. More thinking needed.

**Review by mallowmew — CHANGES_REQUESTED** [review]
* I only have black socks and I cannot add new sock colours.
* I can't add new sizes or patterns either

File: index.php (hunk 1 above)
Comment (Jago971) [reply index.php]: will do

File: src/functions.php (hunk 2 above)
Comment (Jago971) [reply src/functions.php]: Do you mean to change this to $_POST

File: src/functions.php [hunk 4]
```diff
@@ -40,4 +48,57 @@ function displaySocksCollection(PDO $db): string {
...
+
+function createDropdown(string $name) : string {
+    $start = "<select name=\"{$name}\" id=\"{$name}\"><option disabled=\"disabled\" selected=\"selected\">$name</option>";
+    $options = dropDownOptions(getTable($name, connectDB()));
+    $end = '</select>';
+    return $start . $options . $end;
+}
+
+function insertIntoDatabase() {
+    if ($_POST) {
```
Comment (mallowmew) [inline src/functions.php]: I think this 'if' should happen 'outside' this function, gating where you call this function on line 3 of index.php.

I'm not sure that `if ($_POST)`  works as a test that there is actual data in the `$_POST` - these superglobals always exist. You should be doing some sort of check like `if (isset($_POST['some required data field']`

You should also be validating and santising your data before you let it get near your database - are all the fields present you expect to be? Has the user provided data with the right types?

File: src/functions.php [hunk 5]
```diff
@@ -40,4 +48,57 @@ function displaySocksCollection(PDO $db): string {
         echo 'not working';
     }
     return createSockDiv($socksArr);
-}
\ No newline at end of file
+}
+
+function getTable(string $table,PDO $db): array {
+    $query = $db->prepare("SELECT `{$table}` FROM `{$table}s`");
+    $result = $query->execute();
```
Comment (mallowmew) [inline src/functions.php]: This should still use placeholders to put data into the query - you might be careful about using only table names you know, but another developer on your team might not be so careful and could expose this query to user input.

**Review by mallowmew — CHANGES_REQUESTED** [review]

File: src/functions.php [hunk 6]
```diff
@@ -8,36 +8,100 @@ function connectDB(): PDO {
...
+    if ($result) {
+        return $query->fetchAll();
+    } else {
+        throw new Exception('error');
     }
-    return createSockDiv($socksArr);
-}
\ No newline at end of file
+}
+function getPatterns(PDO $db): array {
```
Comment (mallowmew) [inline src/functions.php:73]: I would like a single new line between each function definition

File: index.php [hunk 7]
```diff
@@ -1,5 +1,32 @@
...
+        insertIntoDatabase($data, $db);
+    } else {
+        echo 'not working';
+    }
+}
+//YOU DIDN'T SEE ANYTHING HERE
+
+//if (isset($_POST['sock'])) {
+//    deleteFromDatabase($db);
+//}
```
Comment (mallowmew) [inline index.php]: Remove commented code lines

File: styles.css [hunk 8]
```diff
@@ -262,8 +317,9 @@ form {
 }
 /*-----------------------------------------------------------------sock hover description box*/
```
Comment (mallowmew) [inline styles.css]: Still got these comment separator things

File: styles.css [hunk 9]
```diff
@@ -292,6 +348,34 @@ form {
...
+/*    animation: descriptionBounce 5s linear infinite;*/
+/*    animation-delay: 1.5s;*/
+/*}*/
+/*.bounce4 {*/
+/*    animation: descriptionBounce 5s linear infinite;*/
+/*    animation-delay: 2s;*/
+/*}*/
+
+
+
```
Comment (mallowmew) [inline styles.css]: tidy up please

File: tests/createSockDivTest.php
Comment (mallowmew) [inline tests/createSockDivTest.php:1]: Can you test what happens if you try to make a sock that has incomplete data in the database? e.g. missing size, pattern or color - these are supposed to be compulsory fields right?

Similarly, test to show what happens if missing description - this should be allowed right?

**Review by mallowmew — CHANGES_REQUESTED** [review]
Nicely formatted code apart from the comments that are just a bit of clutter

File: index.php [hunk 10]
```diff
@@ -1,5 +1,36 @@
...
+    }
+}
+//YOU DIDN'T SEE ANYTHING HERE
+
+if (isset($_POST['deletesock'])) {
+    $sock = filter_var($_POST['deletesock'], FILTER_VALIDATE_INT);
+    if ($sock) {
+        $data = ['id' => $sock];
+        deleteFromDatabase($data, $db);
+    }
```
Comment (mallowmew) [inline index.php]: Excess comment, stuff leaking in from another story, remove this code fom this story.

File: src/functions.php [hunk 11]
```diff
@@ -8,36 +8,110 @@ function connectDB(): PDO {
...
+
+function getSizes(PDO $db): array {
+    $query = $db->prepare("SELECT `id`, `name` FROM `sizes`");
+    $result = $query->execute();
+    if ($result) {
+        return $query->fetchAll();
+    } else {
+        throw new Exception('error');
+    }
+}
```
Comment (mallowmew) [inline src/functions.php:61]: I would like a single new line between each function definition

File: src/functions.php [hunk 12]
```diff
@@ -8,36 +8,110 @@ function connectDB(): PDO {
...
-        echo 'not working';
+        throw new Exception('error');
+    }
+    $sizes = getSizes($db);
+    $patterns = getPatterns($db);
+    $colors = getColors($db);
+    return createSockDiv($socksArr, $sizes, $patterns, $colors);
+}
+
+
```
Comment (mallowmew) [inline src/functions.php]: I would like a single new line between each function definition

File: src/functions.php [hunk 13]
```diff
@@ -8,36 +8,110 @@ function connectDB(): PDO {
...
+function insertIntoDatabase(array $sanitizedData, PDO $db) {
+        $query = $db->prepare("INSERT INTO `socks` (`name`, `size`, `pattern`, `color`, `description`) VALUES (:name, :size, :pattern, :color, :description);");
+        $query->execute([
+            'name' => $sanitizedData['name'],
+            'size' => $sanitizedData['size'],
+            'pattern' => $sanitizedData['pattern'],
+            'color' => $sanitizedData['color'],
+            'description' => $sanitizedData['description'] ?? ''
+        ]);
```
Comment (mallowmew) [inline src/functions.php]: Inconsistent indentation with other functions in this file

File: tests/createSockDivTest.php [hunk 14]
```diff
@@ -6,74 +6,193 @@
...
+        $expected .= "<div class=\"description flex\">";
+        $expected .= "<h3>name</h3>";
+        $expected .= "<div class=\"description-container\"><p></p><p class=\"cover\">...</p></div>";
+        $expected .= "</div>";
+        $expected .= "</div>";
+
+        $actual = createSockDiv($input, $sizes, $patterns, $colors);
+        $this->assertEquals($expected, $actual);
+    }
+    public function testCreateSockDivMissingField(): void
```
Comment (mallowmew) [inline tests/createSockDivTest.php:123]: I would like a single new line between each function definition

File: tests/createSockDivTest.php [hunk 15]
```diff
@@ -6,74 +6,193 @@
...
+        ];
+        $sizes = [
+            [
+                'id'=>1,
+                'name'=>'XS'
+            ]
+        ];
+        $patterns = [
+            ['id'=>1,
+                'name'=>'zigzag']
```
Comment (mallowmew) [inline tests/createSockDivTest.php]: inconsistent formatting

File: tests/createSockDivTest.php [hunk 16]
```diff
@@ -6,74 +6,193 @@
...
+        ];
+        $colors = [
+            ['id'=>1,
+                'name'=>'yellow']
+        ];
```
Comment (mallowmew) [inline tests/createSockDivTest.php]: Inconsistent formatting

**Review by mallowmew — CHANGES_REQUESTED** [review]

**Pull request merged**